### DISCUSSION TITLE
How to sort Tailwind classes?

### DISCUSSION THREAD

**Discussion by francoism90** [saadeghi/daisyui#2124] [Q&A]
I'm using VueJS 3 (including TS) and this is my `.eslintrc.js`:
```ts
/* eslint-env node */
require('@rushstack/eslint-patch/modern-module-resolution')

module.exports = {
  root: true,
  extends: [
    'plugin:vue/vue3-recommended',
    'plugin:tailwindcss/recommended',
    '@vue/eslint-config-typescript',
  ],
  parserOptions: {
    ecmaVersion: 'latest'
  },
  settings: {
    tailwindcss: {
      whitelist: [
        'menu\\-(compact)',
      ],
    },
  }
}
``` 
I have the following dev packages installed:
```json
"devDependencies": {
    "@rushstack/eslint-patch": "^1.3.2",
    "@tsconfig/node18": "^18.2.0",
    "@types/jsdom": "^21.1.1",
    "@types/node": "^20.4.1",
    "@types/qs": "^6.9.7",
    "@vitejs/plugin-vue": "^4.2.3",
    "@vitejs/plugin-vue-jsx": "^3.0.1",
    "@vue/eslint-config-typescript": "^11.0.3",
    "@vue/test-utils": "^2.4.0",
    "@vue/tsconfig": "^0.4.0",
    "autoprefixer": "^10.4.14",
    "eslint": "^8.44.0",
    "eslint-plugin-tailwindcss": "^3.13.0",
    "eslint-plugin-vue": "^9.15.1",
    "jsdom": "^22.1.0",
    "npm-run-all": "^4.1.5",
    "postcss": "^8.4.25",
    "tailwindcss": "^3.3.2",
    "typescript": "^5.1.6",
    "vite": "^4.4.2",
    "vite-plugin-pwa": "^0.16.4",
    "vite-svg-loader": "^4.0.0",
    "vitest": "^0.33.0",
    "vue-tsc": "^1.8.4"
  }
```
The problem is the invalid sorting of DaisyUI classes.

Result:
```html
<div class="card-compact card rounded-none"/>
<div class="menu-compact menu"/>
<button type="button" class="btn-secondary btn" />
```
Expected:
```html
<div class="card card-compact rounded-none"/>
<div class="menu menu-compact"/>
<button type="button" class="btn btn-secondary" />
```
I don't know how to solve this, I've tried multiple configurations, but they produce the same results.
Is this a DaisyUI issue or something with Tailwind sort plugin?

Thanks.

**Comment by saadeghi** [comment]
> Is this a DaisyUI issue or something with Tailwind sort plugin?

None of them... 

- daisyUI is just a plugin for Tailwind CSS that adds additional class names to Tailwind. It has nothing to do about sorting or eslint (just like when you use Tailwind CSS without the eslint or prettier plugin. There is no sorting by default)
- eslint plugin for Tailwind CSS is only promised to sort the original Tailwind CSS class names. Not the additional class names provided by plugins.

So both of them are doing their job correctly.  
I can't expect eslint plugin to sort daisyUI as well. And as far as I know there is no eslint plugin for sorting daisyUI class names (I would appreciate if anyone wants to develop a plugin for that)  

And I'm not sure if eslint plugin for Tailwind CSS has a config to sort additional class names. On their docs it says they are using `prettier-plugin-tailwindcss` for sorting and I didn't find any customization config for that either.

If anyone knows a way to sort custom class names, please let us know 🙏

**Reply by francoism90** [reply]
@saadeghi Thanks for your detailed answer, and of course for maintaining DaisyUI. :)

It's weird, sometimes it does use the correct order (card card-compact), however it seems it's get confused and later reverse it back again. This happens in VSCode, when using CLI it always uses the incorrect order. I think it doesn't know what to do, and it's in a loop or something.

I'll ask upstream, maybe it's possible to somehow set a regex, like menu-* classes should always be set after menu.

I'll report back.

**Reply by francoism90** [reply]
@saadeghi https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/181
